src/pdf_beamtime/src/simple_server.cpp

```diff
@@ -0,0 +1,182 @@
+
+  this->move_group_interface_ = new MoveGroupInterface(this->node_, move_group_name);
+
+  // Create the action server
+  this->action_server_ = rclcpp_action::create_server<SimpleActionMsg>(
+    this->node_,
+    this->ACTION_NAME,
+    std::bind(&SimpleServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
+    std::bind(&SimpleServer::handle_cancel, this, std::placeholders::_1),
+    std::bind(&SimpleServer::handle_accepted, this, std::placeholders::_1));
+
+}
+rclcpp::node_interfaces::NodeBaseInterface::SharedPtr SimpleServer::getNodeBaseInterface()
+// Expose the node base interface so that the node can be added to a component manager.
+{
+  return node_->get_node_base_interface();
+}
+
+rclcpp_action::GoalResponse SimpleServer::handle_goal(
+  const rclcpp_action::GoalUUID & uuid,
+  std::shared_ptr<const SimpleActionMsg::Goal> goal)
+{
+  (void)uuid;
+  return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+void SimpleServer::handle_accepted(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<SimpleActionMsg>> goal_handle)
+{
+  using namespace std::placeholders;
+  // this needs to return quickly to avoid blocking the executor, so spin up a new thread
+  std::thread{std::bind(&SimpleServer::execute, this, _1), goal_handle}.detach();
+}
+
+// Receiving the cancel request.
+rclcpp_action::CancelResponse SimpleServer::handle_cancel(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<SimpleActionMsg>> goal_handle)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to cancel goal");
+  (void)goal_handle;
+  return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+void SimpleServer::execute(
+  const std::shared_ptr<rclcpp_action::ServerGoalHandle<SimpleActionMsg>> goal_handle)
+{
+  // A cool Moveit function
+  std::vector<double> home_angles = node_->get_parameter("home_angles").as_double_array();
+
+  this->move_group_interface_->setJointValueTarget(home_angles);
+  // Create a plan to that target pose
+  auto const [success, plan] = [this] {
+      moveit::planning_interface::MoveGroupInterface::Plan msg;
+      auto const ok = static_cast<bool>(this->move_group_interface_->plan(msg));
+      return std::make_pair(ok, msg);
+    }();
+  // Execute the plan
+  if (success) {
+    this->move_group_interface_->execute(plan);
+  } else {
+    RCLCPP_ERROR(node_->get_logger(), "Planning failed!");
+  }
+
+}
+
+std::vector<moveit_msgs::msg::CollisionObject> SimpleServer::create_env()
+// Builds a vector of obstacle as defined in the .yaml file and returns the vector
+{
+  obstacle_type_map.insert(std::pair<std::string, int>("CYLINDER", 1));
+  obstacle_type_map.insert(std::pair<std::string, int>("BOX", 2));
+
+  // int num_objects = node_->get_parameter("num_objects").as_int();
+  std::vector<std::string> object_names = node_->get_parameter("object_names").as_string_array();
+
+  std::vector<moveit_msgs::msg::CollisionObject> all_obstacles;
+
+  // Create objects in a recursion
```

Comment:
Not a recursion, a loop.

Reply:
the word recursion is changed to a loop
https://github.com/maffettone/erobs/blob/9f696b7d2d205fd2038f6a764db896805352b3fb/src/pdf_beamtime/src/pdf_beamtime_server.cpp#L89